builder.go

```diff
@@ -75,110 +90,186 @@ func (b *Builder) Build() (*BuildMetadata, error) {
+			// this allows snapshot buildpacks to have cache and opt to do nothing
+			err := bp.ApplyLayerSnapshots(b.Snapshotter, b.LayersDir)
+			if err != nil {
+				return nil, err
+			}
 		}
 
-		if err := os.MkdirAll(bpPlanDir, 0777); err != nil {
+		launchData, newPlan, bpBOM, err := b.build(bp, appDir, layersDir, platformDir, planDir, plan)
+		if err != nil {
 			return nil, err
 		}
-		bpPlanPath := filepath.Join(bpPlanDir, "plan.toml")
 
-		foundPlan := plan.find(bp.noAPI())
-		if api.MustParse(bp.API).Equal(api.MustParse("0.2")) {
-			for i := range foundPlan.Entries {
-				foundPlan.Entries[i].convertMetadataToVersion()
+		if useSnapshotter {
+			bpDirName := launch.EscapeID(bp.ID)
+			bpLayersDir := filepath.Join(b.LayersDir, bpDirName)
+			layerDir := filepath.Join(bpLayersDir, "snapshot")
+			if err := os.MkdirAll(layerDir, 0777); err != nil {
+				return nil, err
 			}
-		}
-		if err := WriteTOML(bpPlanPath, foundPlan); err != nil {
-			return nil, err
-		}
 
-		cmd := exec.Command(
-			filepath.Join(bpInfo.Path, "bin", "build"),
-			bpLayersDir,
-			platformDir,
-			bpPlanPath,
-		)
-		cmd.Dir = appDir
-		cmd.Stdout = b.Out.Writer()
-		cmd.Stderr = b.Err.Writer()
-
-		if bpInfo.Buildpack.ClearEnv {
-			cmd.Env = b.Env.List()
-		} else {
-			cmd.Env, err = b.Env.WithPlatform(platformDir)
-			if err != nil {
+			snapshotPath := filepath.Join(layerDir, "snapshot.tgz")
+			if err := b.Snapshotter.TakeSnapshot(snapshotPath); err != nil {
 				return nil, err
 			}
+			layerTOMLPath := filepath.Join(bpLayersDir, "snapshot.toml")
+			md := BuildpackLayerMetadataFile{
+				Cache:  true,
+				Build:  true,
+				Launch: false,
+			}
+			if err := WriteTOML(layerTOMLPath, md); err != nil {
+				return nil, cmd.FailErr(err, "write build layer metadata")
+			}
 		}
-		cmd.Env = append(cmd.Env, EnvBuildpackDir+"="+bpInfo.Path)
 
-		if err := cmd.Run(); err != nil {
-			return nil, NewLifecycleError(err, ErrTypeBuildpack)
-		}
-		if err := setupEnv(b.Env, bpLayersDir); err != nil {
-			return nil, err
-		}
-		var bpPlanOut BuildpackPlan
-		if _, err := toml.DecodeFile(bpPlanPath, &bpPlanOut); err != nil {
-			return nil, err
-		}
-		var bpBOM []BOMEntry
-		plan, bpBOM = plan.filter(bp, bpPlanOut)
+		plan = newPlan
 		bom = append(bom, bpBOM...)
+		procMap.add(launchData.Processes)
+		slices = append(slices, launchData.Slices...)
+		labels = append(labels, launchData.Labels...)
+	}
+
+	if err := b.convertMetadataToVersion(bom); err != nil {
+		return nil, err
+	}
+
+	return &BuildMetadata{
+		BOM:        bom,
+		Buildpacks: b.Group.Group,
+		Labels:     labels,
+		Processes:  procMap.list(),
+		Slices:     slices,
+	}, nil
+}
 
-		var launch LaunchTOML
-		tomlPath := filepath.Join(bpLayersDir, "launch.toml")
-		if _, err := toml.DecodeFile(tomlPath, &launch); os.IsNotExist(err) {
+func (bp Buildpack) ApplyLayerSnapshots(snapshotter LayerSnapshotter, layersDir string) error {
```

Comment:
This function name is odd to me because while there may be multiple layers, we only expect to apply one snapshot... maybe `ApplyLayersSnapshot`? Also just for my understanding, when would you expect a stack buildpack to have multiple layers? I thought it could only ever create one layer (the snapshot)...

Reply:
I left that intentionally open. We could give the stackpack developer a way to potentially create multiple layers. I was imagining a situation where instead of generating a single `layer.toml` for a `snapshot` layer. The buildpack developer could instead be responsible for creating `layer.toml` but with different `exclude` or `include` paths (along with `cache=false` potentially. Then we could split the kaniko generated snapshot into X distinct tars. This might be useful for some advanced cache scenarios where you don't want to restore or store _all_ of the snapshot on subsequent re-runs.

Reply:
Ohh right... I recall that was part of the RFC that is unimplemented for now.